### ISSUE TITLE
Improving feedback when Jupyter fails to load the pipeline

### ISSUE THREAD

**Issue by edublancas** [ploomber/ploomber#896]
[Our Jupyter integration](https://docs.ploomber.io/en/latest/get-started/basic-concepts.html#the-cell-injection-process) requires loading the user's pipeline whenever they open a notebook ([source code](https://github.com/ploomber/ploomber/blob/master/src/ploomber/jupyter/manager.py#L104)). However, the loading process will fail if the `pipeline.yaml` has issues. For example, if the `pipeline.yaml` is an invalid YAML, if the user is not following the schema or if they're declaring a script as a task but the file doesn't exist - there are many reasons why a pipeline wouldn't load successfully.

Currently, if the pipeline loading process file, it will do so silently and the [cell injection](https://docs.ploomber.io/en/latest/get-started/basic-concepts.html#the-cell-injection-process) process won't work; the only way for users to know what happened is to take a look at the Jupyter console (the one where the `jupyter lab` command is executed). However, looking at the console is not common practice and many users will think that the extension is broken.

So we need to provide some feedback when the pipeline fails to load in the main Jupyter interface in a non-obtrusive way (a visible message that does not require the user to click anything to dismiss it).

The challenge here is to find how to communicate the process that's running Jupyter (which is where our plugin runs) with the frontend so we can display an alert. I did some quick research but couldn't find anything so more research is needed.

Note that this will involve adding a frontend extension to Ploomber (https://github.com/ploomber/ploomber/issues/893 also needs that so hopefully by the time we start working on this, that other issue has made some progress and the basic setup for the extension is already done).

Here's the documentation to [build extensions](https://jupyterlab.readthedocs.io/en/stable/extension/extension_dev.html#), here's a [tutorial](https://jupyterlab.readthedocs.io/en/stable/extension/extension_tutorial.html) that shows how to build a simple extension, and here's a list of [examples.](https://github.com/jupyterlab/extension-examples)

**Comment by edublancas** [comment]
I ran a little experiment using [ipylab](https://github.com/jtpio/ipylab) to see if running it in our [plugin](https://github.com/ploomber/ploomber/blob/master/src/ploomber/jupyter/manager.py) would work, but it didn't. I also could not find any references online, so I looks like our best resource is to get help in the Jupyter discourse

**Comment by Wxl19980214** [comment]
Can I get more information on this issue? I was running one example and I changed pipeline.yaml to invalid format and use a scripts that doesnt exist, i get following error messages:

```
(soorgeon) ➜  exploratory git:(main) ✗ ploomber build
Loading pipeline...
There's a new Ploomber version available (0.19.9), you're running 0.19.6. To upgrade: pip install ploomber --upgrade
Error: Failed to initialize '/Users/xilinwang/Documents/ploomber/soorgeon/examples/exploratory/tasks/jkhkjhkj.ipynb'. File does not exist.
To create it, run: ploomber scaffold
(soorgeon) ➜  exploratory git:(main) ✗ ploomber build
Loading pipeline...
Error: Failed to initialize spec. Got invalid YAML
while parsing a block mapping
  in "<unicode string>", line 1, column 1:
    tasks:
    ^
expected <block end>, but found '-'
  in "<unicode string>", line 8, column 1:
    - source: tasks/clean.ipynb

```

Isnt there's enough feedback? I am not sure how to reproduce the situation?

**Comment by Wxl19980214** [comment]
`Our Jupyter integration`  what is this? Like convert notebook to py file? I am a little confused. 
I thought the use case of ploomber for notebook user is that all they have is a notebook.ipynb, and we asks them to write a pipeline.yaml, and we make some change to their notebook.ipynb , like injecting cells for upstream and products? I don't really understand this:

[Our Jupyter integration](https://docs.ploomber.io/en/latest/get-started/basic-concepts.html#the-cell-injection-process) requires loading the user's pipeline whenever they open a notebook ([source code](https://github.com/ploomber/ploomber/blob/master/src/ploomber/jupyter/manager.py#L104)).

Why does user open a notebook has anything to do with ploomber?

**Comment by Wxl19980214** [comment]
@edublancas I think this is not achievable just by writing code in server side. We have to communicate between server and client and thus client must have some codes to receive the message we sent from server. I think we might need to implement a client side extension.

**Comment by edublancas** [comment]
have you found any information on how the backend communicates with the client-side extension?

**Comment by Wxl19980214** [comment]
I am not a jupyter notebook user so I have no idea how it works. The closet thing I found is the [comm](https://jupyter-notebook.readthedocs.io/en/stable/comms.html) object. It looks like it is communicating between kernel and front end. But i don't know if kernel is the same as backend, i feel like it is. 

But no matter what, we have to have something on client side.

**Comment by Wxl19980214** [comment]
I am going to try something today and tomorrow. I will follow the tutorial and create a client side extension to see if it works.

**Comment by edublancas** [comment]
awesome thanks for the update

**Comment by idomic** [comment]
@Wxl19980214 any updates here?

**Comment by edublancas** [comment]
I discussed this with @Wxl19980214 a week ago. and share what I found when I first tried it, putting it here:

we thought that modifying the notebook and adding one cell with an error is one option. because Xilin couldn't
find a way to communicate the backend with the frontend. 

Xilin: what's the conclusion here? I think injecting a cell is our very last resort since It'll be confusing to users. If we cannot communicate the backend with the frontend. maybe we can create a file in ~/.ploomber from the backend and the frontend can pick it up to display the error? The main challenge here is that doing this requires writing a Jupyter extension. @e1ha  is working on this as part of the autocomplete, so I suggest merging a PR that adds the skeleton of a Ploomber frontend extension so we have all the elements in place for both the autocomplete and this.

I tried the "injecting a cell" approach and didn't like it, here are my findings just for reference:

---

>I did some quick experiments in the [jupyter branch](https://github.com/ploomber/ploomber/tree/jupyter).
> If you check the jupyter branch, you'll see that the code works - it adds a new cell if the pipeline fails to load. the problem is that the [tests dont pass](https://github.com/ploomber/ploomber/blob/0c7f7ac5a2a03b93402caabc00f473a54e9f15c3/tests/test_jupyter.py#L30). I realized that the problem is that Jupyter has a security model, so if we add a new cell, it breaks. What I don't recall is why adding this markdown cell breaks them but our current step of adding a code cell doesn't. perhaps it only affects markdown cells.

### Add a cell showing the error message and asking to check the Jupyter logs.
​
The initial tests for the contents manager break because our contents manager modifies `.get`, and adds a new cell to the notebooks. The best solution is to alter the `notary.sign` to ignore the cells that we add.
​
However, this introduces a problem because if someone takes a trusted notebook, adds the cells with the tags, and adds malicious code, the signature will match and the tampered notebook will be trusted.
​
To ensure this doesn't happen, the get method should clean up any existing cells with the given tags, so they never make it to the user's frontend.
​
## Files
​
[test_manager](https://github.com/jupyter/notebook/blob/master/notebook/services/contents/tests/test_manager.py)
​
[large file manager](https://github.com/jupyter/notebook/blob/ccd9665571107e02a325a738b8baebd6532b2d3d/notebook/services/contents/largefilemanager.py)
​
[contents manager](https://github.com/jupyter/notebook/blob/ccd9665571107e02a325a738b8baebd6532b2d3d/notebook/services/contents/manager.py#L479)
​
[file manager (original .get implementation)](https://github.com/jupyter/notebook/blob/ccd9665571107e02a325a738b8baebd6532b2d3d/notebook/services/contents/filemanager.py#L405)
​
​
[security model](https://jupyter-notebook.readthedocs.io/en/stable/security.html#notebook-security)
​
​
[notebook notary](https://github.com/jupyter/nbformat/blob/f101cd2b062609823e9cdbf0e8de697a6c22638e/nbformat/sign.py#L312)

**Comment by Wxl19980214** [comment]
I am just going to share some of my findings.  

1st possible approach is the comm object. This [one](https://github.com/jupyterlab/jupyterlab/blob/master/packages/services/examples/browser/src/comm.ts) looks promising. The problem is that when I tried it, I couldn't find a place to implement the backend part of comm object. I also couldn't find kernel information in our python code. But now I found this [function](https://github.com/ploomber/ploomber/blob/master/src/ploomber/sources/notebooksource.py#L760) might help.

2nd approach is to somehow raise an error that can be propagate to front end and then catch it in jupyter client. 

3rd approach is to inject the cell, and yeah it works, but it breaks the tests because all markdown cells are not trusted.

Here is the [post](https://discourse.jupyter.org/t/sending-error-message-from-server-to-user-interface/15204/6) in jupyter forum, I hope it contains some useful information. 

Bottom line is we do need to implement a typescript front end to receive the message anyway and that was a big problem for me.

**Comment by edublancas** [comment]
thanks for diving into this @Wxl19980214; this information is really useful! I think let's put this on hold. @idomic and I will sync to decide how we go about this
